internal/subscriber/retry/delayconsumer.go

```diff
@@ -0,0 +1,186 @@
+		return nil, err
+	}
+
+	// on init, make sure to call resume. This is done just to ensure any previously paused consumers get resumed on boot up.
+	consumer.Resume(ctx, messagebroker.ResumeOnTopicRequest{Topic: topic})
+
+	return &DelayConsumer{
+		ctx:        delayCtx,
+		cancelFunc: cancel,
+		topic:      topic,
+		consumer:   consumer,
+		subs:       subs,
+		bs:         bs,
+		handler:    handler,
+		doneCh:     make(chan struct{}),
+	}, nil
+}
+
+// Run spawns the delay-consumer
+func (dc *DelayConsumer) Run(ctx context.Context) {
+	defer close(dc.doneCh)
+
+	logger.Ctx(ctx).Infow("delay-consumer: running", dc.LogFields()...)
+	for {
+		select {
+		case <-dc.ctx.Done():
+			logger.Ctx(dc.ctx).Infow("delay-consumer: stopping <-ctx.Done() called", dc.LogFields()...)
+			dc.bs.RemoveConsumer(ctx, messagebroker.ConsumerClientOptions{GroupID: dc.subs.GetDelayConsumerGroupID(dc.topic), GroupInstanceID: dc.subs.GetDelayConsumerGroupInstanceID(dc.topic)})
+			dc.consumer.Close(dc.ctx)
+			return
+		default:
+			if dc.cachedMsg != nil && dc.cachedMsg.CanProcessMessage() {
+				dc.resume()
+			}
+
+			resp, err := dc.consumer.ReceiveMessages(dc.ctx, messagebroker.GetMessagesFromTopicRequest{NumOfMessages: 10, TimeoutMs: int(defaultBrokerOperationsTimeoutMs)})
+			if err != nil {
+				if !messagebroker.IsErrorRecoverable(err) {
+					logger.Ctx(dc.ctx).Errorw("delay-consumer: error in receiving messages", dc.LogFields("error", err.Error())...)
+					return
+				}
+			}
+
+			for _, msg := range resp.PartitionOffsetWithMessages {
+				dc.processMsg(msg)
+			}
+		}
+	}
+}
+
+// resumes a paused delay-consumer. Additionally process any previously cached messages in buffer.
+func (dc *DelayConsumer) resume() {
+	logger.Ctx(dc.ctx).Infow("delay-consumer: resuming", dc.LogFields("error", nil)...)
+	dc.consumer.Resume(dc.ctx, messagebroker.ResumeOnTopicRequest{Topic: dc.topic})
+	dc.isPaused = false
+
+	// process the cached message as well
+	dc.processMsg(*dc.cachedMsg)
+	dc.cachedMsg = nil
+}
+
+// paused an active delay-consumer. Additionally caches the last seen message locally.
+func (dc *DelayConsumer) pause(msg *messagebroker.ReceivedMessage) {
+	logger.Ctx(dc.ctx).Infow("delay-consumer: pausing", dc.LogFields()...)
+	dc.consumer.Pause(dc.ctx, messagebroker.PauseOnTopicRequest{Topic: dc.topic})
+	dc.isPaused = true
+	dc.cachedMsg = msg
+}
+
+// push a message onto the configured dead letter topic
+func (dc *DelayConsumer) pushToDeadLetter(msg *messagebroker.ReceivedMessage) error {
+	logger.Ctx(dc.ctx).Infow("delay-consumer: pushing to dead-letter", dc.LogFields()...)
+	dlProducer, err := dc.bs.GetProducer(dc.ctx, messagebroker.ProducerClientOptions{
+		Topic:     msg.DeadLetterTopic,
+		TimeoutMs: defaultBrokerOperationsTimeoutMs,
+	})
+	if err != nil {
+		return err
+	}
+
+	_, err = dlProducer.SendMessage(dc.ctx, messagebroker.SendMessageToTopicRequest{
+		Topic:         msg.DeadLetterTopic,
+		Message:       msg.Data,
```

Comment:
we should add some attributes to the message, pubsub adds few attributes related to source topic details in the deadletter topic

Reply:
Agreed! Wasn't sure of the fields to be published in the outgoing message. Can you share a sample gPubSub message with the needed fields for reference?

Reply:
I see that [delivery_attempts](https://github.com/razorpay/metro/pull/245) has already been added by @razorao. Anymore needed?